File: src/lib/time.ts
import type { ReminderWindow } from "./types";

export function minutesFromTime(value: string): number | null {
  if (!/^\d{2}:\d{2}$/.test(value)) return null;
  const [hour, minute] = value.split(":").map(Number);
  if (hour > 23 || minute > 59) return null;
  return hour * 60 + minute;
}

export function validateReminderWindow(window: ReminderWindow): string[] {
  const errors: string[] = [];
  const start = minutesFromTime(window.startTime);
  const end = minutesFromTime(window.endTime);
  if (!window.name.trim()) errors.push("Name is required");
  if (start === null) errors.push("Start time format is invalid");
  if (end === null) errors.push("End time format is invalid");
  if (start !== null && end !== null && start >= end) {
    errors.push("Start time must be earlier than end time");
  }
  if (!Number.isFinite(window.remindIntervalMinutes) || window.remindIntervalMinutes < 1) {
    errors.push("Reminder interval must be at least 1 minute");
  }
  return errors;
}

export function findOverlaps(windows: ReminderWindow[]): string[] {
  const overlaps: string[] = [];
  const enabled = windows.filter((window) => window.enabled);
  for (let leftIndex = 0; leftIndex < enabled.length; leftIndex += 1) {
    const left = enabled[leftIndex];
    const leftStart = minutesFromTime(left.startTime);
    const leftEnd = minutesFromTime(left.endTime);
    if (leftStart === null || leftEnd === null) continue;
    for (const right of enabled.slice(leftIndex + 1)) {
      const rightStart = minutesFromTime(right.startTime);
      const rightEnd = minutesFromTime(right.endTime);
      if (rightStart === null || rightEnd === null) continue;
      if (leftStart < rightEnd && rightStart < leftEnd) {
        overlaps.push(`${left.name} / ${right.name}`);
      }
    }
  }
  return overlaps;
}

File: src-tauri/src/time.rs
use crate::config::ReminderWindow;

pub fn minutes_from_time(value: &str) -> Result<u32, String> {
    let (hour, minute) = value
        .split_once(':')
        .ok_or_else(|| format!("invalid time: {value}"))?;
    if hour.len() != 2 || minute.len() != 2 {
        return Err(format!("invalid time: {value}"));
    }
    let hour: u32 = hour
        .parse()
        .map_err(|_| format!("invalid hour in time: {value}"))?;
    let minute: u32 = minute
        .parse()
        .map_err(|_| format!("invalid minute in time: {value}"))?;
    if hour > 23 || minute > 59 {
        return Err(format!("invalid time: {value}"));
    }
    Ok(hour * 60 + minute)
}

pub fn time_in_window(current: &str, start: &str, end: &str) -> Result<bool, String> {
    let current = minutes_from_time(current)?;
    let start = minutes_from_time(start)?;
    let end = minutes_from_time(end)?;
    Ok(start < end && current >= start && current < end)
}

pub fn validate_window(start: &str, end: &str, interval_minutes: u32) -> Result<(), String> {
    let start = minutes_from_time(start)?;
    let end = minutes_from_time(end)?;
    if start >= end {
        return Err("startTime must be earlier than endTime".to_string());
    }
    if interval_minutes < 1 {
        return Err("remindIntervalMinutes must be at least 1".to_string());
    }
    Ok(())
}

pub fn find_overlaps(windows: &[ReminderWindow]) -> Vec<(String, String)> {
    let mut overlaps = Vec::new();
    for (index, left) in windows.iter().enumerate() {
        if !left.enabled {
            continue;
        }
        let Ok(left_start) = minutes_from_time(&left.start_time) else {
            continue;
        };
        let Ok(left_end) = minutes_from_time(&left.end_time) else {
            continue;
        };
        for right in windows.iter().skip(index + 1) {
            if !right.enabled {
                continue;
            }
            let Ok(right_start) = minutes_from_time(&right.start_time) else {
                continue;
            };
            let Ok(right_end) = minutes_from_time(&right.end_time) else {
                continue;
            };
            if left_start < right_end && right_start < left_end {
                overlaps.push((left.id.clone(), right.id.clone()));
            }
        }
    }
    overlaps
}

pub fn add_minutes(time: &str, minutes: u32) -> Result<String, String> {
    let total = minutes_from_time(time)? + minutes;
    Ok(format!("{:02}:{:02}", (total / 60) % 24, total % 60))
}

pub fn is_at_or_after(current: &str, target: &str) -> Result<bool, String> {
    Ok(minutes_from_time(current)? >= minutes_from_time(target)?)
}
